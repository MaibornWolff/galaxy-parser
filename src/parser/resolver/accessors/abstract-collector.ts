import { debuglog, type DebugLoggerFunction } from "node:util";
import { type QueryCapture } from "tree-sitter";
import { type ParsedFile } from "../../metrics/metric.js";
import { formatCaptures } from "../../../helper/helper.js";
import { QueryBuilder } from "../../queries/query-builder.js";
import { type TypeInfo } from "../types/abstract-collector.js";
import { SimpleQueryStatement } from "../../queries/query-statements.js";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export type Accessor = {
    name: string;
    FQTNInfos: TypeInfo[];
    filePath: string;
    returnType: string;
};

export abstract class AbstractCollector {
    getPublicAccessorsFromFile(
        parsedFile: ParsedFile,
        FQTNsFromFile: Map<string, TypeInfo>,
    ): Map<string, Accessor[]> {
        const publicAccessorsMap = new Map<string, Accessor[]>();

        if (this.getAccessorsQuery()) {
            const { filePath, language, tree } = parsedFile;

            const queryBuilder = new QueryBuilder(language);
            queryBuilder.setStatements([new SimpleQueryStatement(this.getAccessorsQuery())]);

            const publicAccessorsQuery = queryBuilder.build();
            let publicAccessorsCaptures: QueryCapture[] = [];
            if (publicAccessorsQuery !== undefined) {
                publicAccessorsCaptures = publicAccessorsQuery.captures(tree.rootNode);
            }

            const accessorsTextCaptures: Array<{
                name: string;
                text: string;
                usageType?: string;
                source?: string;
            }> = formatCaptures(tree, publicAccessorsCaptures);

            dlog("public accessors captures", accessorsTextCaptures);

            // First index must be the return type
            // Second index must be the accessor name
            for (let index = 0; index < accessorsTextCaptures.length; index += 1) {
                const publicAccessor: Accessor = {
                    name: "",
                    FQTNInfos: [...FQTNsFromFile.values()],
                    filePath,
                    returnType: accessorsTextCaptures[index].text,
                };

                index++;
                publicAccessor.name = accessorsTextCaptures[index].text;

                if (publicAccessorsMap.get(publicAccessor.name) === undefined) {
                    publicAccessorsMap.set(publicAccessor.name, []);
                }

                // TODO prevent duplicate adds by checking already existing combinations
                //  of name, returnType, and filepath
                publicAccessorsMap.get(publicAccessor.name)?.push(publicAccessor);
            }
        }

        return publicAccessorsMap;
    }

    protected abstract getAccessorsQuery(): string;
}
