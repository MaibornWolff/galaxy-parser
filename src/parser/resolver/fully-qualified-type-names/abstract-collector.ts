import { type ParsedFile } from "../../metrics/metric.js";
import { QueryStrategy } from "./resolver-strategy/query-strategy.js";
import { FilenameStrategy } from "./resolver-strategy/filename-resolver.js";

export type FQTNInfo = {
    namespace: string;
    className: string;
    classType: "interface" | "class";
    source: string;
    namespaceDelimiter: string;
    extendedClass?: string;
    implementedClasses: string[];
};

export enum NamespaceResolvingStrategy {
    Query,
    Filename,
}

export abstract class AbstractCollector {
    getFQTNsFromFile(parsedFile: ParsedFile): Map<string, FQTNInfo> {
        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Query) {
            return new QueryStrategy().getFQTNsFromFile(
                parsedFile,
                this.getNamespaceDelimiter(),
                this.getNamespacesQuery(),
            );
        }

        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Filename) {
            return new FilenameStrategy().getFullyQTNs();
        }

        throw new Error(
            "Unsupported Namespace Resolving Strategy " +
                this.getNamespaceResolvingStrategy().toString(),
        );
    }

    protected abstract getNamespaceResolvingStrategy(): NamespaceResolvingStrategy;
    protected abstract getNamespaceDelimiter(): string;
    protected abstract getNamespacesQuery(): string;
}
