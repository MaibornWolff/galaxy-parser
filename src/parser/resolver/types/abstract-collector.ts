import { type Query, type SyntaxNode } from "tree-sitter";
import { type ParsedFile } from "../../metrics/metric.js";
import { TypesQueryStrategy } from "./resolver-strategy/types-query-strategy.js";
import { FileNameStrategy } from "./resolver-strategy/filename-resolver.js";

type TypeName = string;
export type TypeInfo = {
    node?: SyntaxNode;
    namespace: string;
    typeName: string;
    classType: ClassType;
    sourceFile: string;
    namespaceDelimiter: string;
    extendedFrom?: string;
    implementedFrom: string[];
};
export type ClassType = "interface" | "class";
export type TypesResolvingStrategy = "Query" | "Filename";

export abstract class AbstractCollector {
    getTypesFromFile(parsedFile: ParsedFile): Map<TypeName, TypeInfo> {
        if (this.getTypesResolvingStrategy() === "Query") {
            return new TypesQueryStrategy().getTypesFromFile(
                parsedFile,
                this.getNamespaceDelimiter(),
                this.getTypesQuery(),
            );
        }

        if (this.getTypesResolvingStrategy() === "Filename") {
            return new FileNameStrategy().getTypes();
        }

        throw new Error("Unsupported Types Resolving Strategy " + this.getTypesResolvingStrategy());
    }
    public abstract getTypesQuery(): Query;
    protected abstract getTypesResolvingStrategy(): TypesResolvingStrategy;
    protected abstract getNamespaceDelimiter(): string;
}
