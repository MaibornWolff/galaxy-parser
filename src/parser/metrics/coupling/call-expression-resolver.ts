import { debuglog, type DebugLoggerFunction } from "node:util";
import { type Relationship } from "../metric.js";
import { type Accessor } from "../../resolver/accessors/abstract-collector.js";
import { type CallExpression } from "../../resolver/call-expressions/abstract-collector.js";
import { type TypeInfo } from "../../resolver/types/abstract-collector.js";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export function getAdditionalRelationships(
    tree: Map<string, Relationship[]>,
    unresolvedCallExpressions: Map<string, CallExpression[]>,
    accessors: Map<string, Accessor[]>,
    alreadyAddedRelationships: Set<string>,
): Relationship[] {
    const additionalRelationships: Relationship[] = [];

    for (const [filePath, callExpressions] of unresolvedCallExpressions) {
        const fileDependencies = tree.get(filePath) ?? [];

        dlog("RESOLVING:", fileDependencies);

        const processedPublicAccessors = new Set<string>();

        const fileAdditionalRelationships: Relationship[] = [];

        for (const callExpression of callExpressions) {
            if (processedPublicAccessors.has(callExpression.name)) {
                continue;
            }

            processedPublicAccessors.add(callExpression.name);

            const qualifiedNameParts = callExpression.name.split(callExpression.namespaceDelimiter);
            if (callExpression.variableNameIncluded) {
                qualifiedNameParts.shift();
            }

            for (let namePart of qualifiedNameParts) {
                // Remove save call character
                if (namePart.endsWith("?")) {
                    namePart = namePart.slice(0, Math.max(0, namePart.length - 1));
                }

                const publicAccessorsOfPrefix = accessors.get(namePart);
                if (publicAccessorsOfPrefix === undefined) {
                    continue;
                }

                const clonedAccessors = [...publicAccessorsOfPrefix];
                dlog(
                    "CLONED_ACCESSORS",
                    namePart,
                    filePath,
                    clonedAccessors.length,
                    fileDependencies,
                );
                dlog(clonedAccessors.toString());

                let added;

                for (const accessor of clonedAccessors) {
                    if (accessor.fromTypes.length === 0) {
                        continue;
                    }

                    // Warning in case of Accessor Conflicts:
                    // In case that multiple accessors would fit the call expression,
                    // add both of them to not lose the correct dependency.
                    // this might lead to higher coupling values,
                    // but it is not that crucial as shown in the master thesis related to MetricGardener
                    for (const type of accessor.fromTypes) {
                        const fullyQualifiedNameCandidate =
                            type.namespace + type.namespaceDelimiter + type.typeName;

                        dlog("\n\n", accessor, " -- ", fullyQualifiedNameCandidate);

                        // FirstAccessor is a property or method
                        // The type of myVariable must be an already added dependency of the current base type/class (filePath),
                        // so that subsequent method calls or attribute accesses can be resolved.
                        // Example:
                        // myVariable.FirstAccessor.SecondAccessor.ThirdAccessor
                        const baseDependency = fileDependencies.find((dependency) => {
                            return dependency.toFQTN === fullyQualifiedNameCandidate;
                        });

                        // In case of chained accesses, look in dependencies added for previous chain elements:
                        // e.g. look up already added dependency of return type of FirstAccessor to resolve SecondAccessor
                        // myVariable.FirstAccessor.SecondAccessor.ThirdAccessor
                        const callExpressionDependency = fileAdditionalRelationships.find(
                            (dependency) => {
                                return dependency.toFQTN === fullyQualifiedNameCandidate;
                            },
                        );

                        if (baseDependency !== undefined) {
                            added = resolveAccessorReturnType(
                                baseDependency,
                                accessor,
                                type,
                                tree,
                                fileAdditionalRelationships,
                                alreadyAddedRelationships,
                            );
                        } else if (callExpressionDependency !== undefined) {
                            added = resolveAccessorReturnType(
                                callExpressionDependency,
                                accessor,
                                type,
                                tree,
                                fileAdditionalRelationships,
                                alreadyAddedRelationships,
                            );
                        }

                        if (added) {
                            // Iterate only for the first namespace of an added accessor:
                            // One namespace of the accessor is fitting and
                            // the corresponding accessor was added as a dependency
                            // It seems not necessary to check the other namespaces of the accessor
                            break;
                        }
                    }
                }
            }
        }

        additionalRelationships.push(...fileAdditionalRelationships);
    }

    return additionalRelationships;
}

function resolveAccessorReturnType(
    matchingDependency: Relationship,
    accessor: Accessor,
    fromType: TypeInfo,
    tree: Map<string, Relationship[]>,
    additionalRelationships: Relationship[],
    alreadyAddedRelationships: Set<string>,
): number {
    // TODO resolve return type (generics, etc.)
    dlog(
        "Lookup Status: ",
        Boolean(matchingDependency),
        " -> check return type add: ",
        accessor.returnType,
        matchingDependency,
        "\n\n",
    );

    const accessorFileDependencies = tree.get(fromType.sourceFile) ?? [];
    dlog("namespace.source", fromType.sourceFile);
    dlog("accessorFileDependencies", accessorFileDependencies);
    for (const accessorFileDependency of accessorFileDependencies) {
        // TODO Imagine that returnType is MyTypeNumberOne
        //  and toTypeName MyType
        //  This would lead to a wrong dependency
        // Substring search using includes() is used because Return Types can look very differently:
        // Collection<MyTypeNumberOne>
        // Map<string, MyTypeNumberOne>
        // etc.
        if (accessor.returnType.includes(accessorFileDependency.toTypeName)) {
            const uniqueId = accessorFileDependency.toFQTN + matchingDependency.fromFQTN;

            if (alreadyAddedRelationships.has(uniqueId)) {
                dlog(
                    "SKIP ADDING RETURN TYPE: ",
                    accessor.returnType,
                    "already added: ",
                    accessorFileDependency,
                    alreadyAddedRelationships.has(uniqueId),
                    "\n\n",
                );
                continue;
            }

            if (matchingDependency.fromFQTN !== accessorFileDependency.toFQTN) {
                alreadyAddedRelationships.add(uniqueId);

                const dependencyClone: Relationship = {
                    fromFQTN: matchingDependency.fromFQTN,
                    fromFile: matchingDependency.fromFile,
                    toTypeName: accessorFileDependency.toTypeName,
                    fromTypeName: accessorFileDependency.toTypeName,
                    toFQTN: accessorFileDependency.toFQTN,
                    toFile: accessorFileDependency.toFile,
                    usageType: "usage",
                };

                dlog("ADD RETURN TYPE: ", accessor.returnType, dependencyClone, "\n\n");
                return additionalRelationships.push(dependencyClone);
            }

            return 0;
        }
    }

    return 0;
}
