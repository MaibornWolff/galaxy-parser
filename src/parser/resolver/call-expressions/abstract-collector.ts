import { debuglog, type DebugLoggerFunction } from "node:util";
import { type QueryCapture, type QueryMatch } from "tree-sitter";
import { QueryBuilder } from "../../queries/query-builder.js";
import { getNameAndTextFromCaptures } from "../../../helper/helper.js";
import { type TypeCollector } from "../type-collector.js";
import { type ParsedFile, type UsageType } from "../../metrics/metric.js";
import { type QueryStatement, SimpleQueryStatement } from "../../queries/query-statements.js";
import { type TypeInfo } from "../types/abstract-collector.js";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export type ImportReference = {
    referenceName: string;
    referenceSuffix: string;
    sourceOfUsing: string;
    alias: string;
    source: string;
    usageType: UsageType;
};

export type TypeUsageCandidate = {
    usedNamespace: string;
    fromNamespace: string;
    sourceOfUsing: string;
    usageType: UsageType;
};

export type UnresolvedCallExpression = {
    name: string;
    variableNameIncluded: boolean;
    namespaceDelimiter: string;
};

export type ImportCapture = {
    name: string;
    text: string;
};
type ImportMatch = {
    importCaptures: ImportCapture[];
};

export abstract class AbstractCollector {
    private readonly fileToImportsBySuffixOrAliasMap = new Map<
        string,
        Map<string, ImportReference>
    >();

    getUsageCandidates(
        parsedFile: ParsedFile,
        typesFromFile: Map<string, TypeInfo>,
    ): {
        candidates: TypeUsageCandidate[];
        unresolvedCallExpressions: UnresolvedCallExpression[];
    } {
        const { filePath } = parsedFile;
        this.fileToImportsBySuffixOrAliasMap.set(filePath, new Map());

        const importReferences = this.getImports(parsedFile);
        if (this.getGroupedImportsQuery().length > 0) {
            importReferences.push(...this.getGroupedImports(parsedFile));
        }

        dlog("Alias Map:", filePath, this.fileToImportsBySuffixOrAliasMap);
        dlog("Import References", filePath, importReferences);

        const { candidates, unresolvedCallExpressions } = this.getUsages(
            parsedFile,
            typesFromFile,
            importReferences,
        );
        dlog("UsagesAndCandidates", filePath, candidates);

        return {
            candidates,
            unresolvedCallExpressions,
        };
    }

    // In C# for example,
    // it is not possible to know which usage belongs to which import
    // Imagine:
    // using System;
    // Console.WriteLine("foo");
    // You cannot be sure that the Console.WriteLine (partial) namespace
    // belongs to the used namespace "System"
    protected abstract indirectNamespaceReferencing(): boolean;
    protected abstract noImportForClassesInSameOrParentNamespaces(): boolean;
    protected abstract getFunctionCallDelimiter(): string;
    protected abstract getNamespaceDelimiter(): string;
    protected abstract getImportsQuery(): QueryStatement;
    protected abstract getGroupedImportsQuery(): string;
    protected abstract getUsagesQuery(): string;

    private getImports(parsedFile: ParsedFile): ImportReference[] {
        const { filePath, language, tree } = parsedFile;

        const queryBuilder = new QueryBuilder(language);
        queryBuilder.setStatements([this.getImportsQuery()]);

        const importsQuery = queryBuilder.build();
        let queryMatches: QueryMatch[] = [];
        if (importsQuery !== undefined) {
            queryMatches = importsQuery.matches(tree.rootNode);
        }

        const importMatches = queryMatches.map((queryMatch): ImportMatch => {
            return {
                importCaptures: getNameAndTextFromCaptures(queryMatch.captures),
            };
        });

        dlog(importMatches.toString());

        return this.buildImportReferences(importMatches, filePath);
    }

    private buildImportReferences(
        importMatches: ImportMatch[],
        filePath: string,
    ): ImportReference[] {
        const importsInFile: ImportReference[] = [];

        for (const importMatch of importMatches) {
            const importReference = this.importMatchToImportReference(importMatch, filePath);
            this.fileToImportsBySuffixOrAliasMap
                .get(filePath)
                ?.set(
                    importReference.alias.length > 0
                        ? importReference.alias
                        : importReference.referenceSuffix,
                    importReference,
                );
            importsInFile.push(importReference);
        }

        return importsInFile;
    }

    private importMatchToImportReference(
        importMatch: ImportMatch,
        filePath: string,
    ): ImportReference {
        const { importCaptures } = importMatch;

        let importSpecifierCapture;
        let alias = "";
        for (const capture of importCaptures) {
            if (capture.name === "alias") {
                alias = capture.text;
            } else if (capture.name === "import_specifier") {
                importSpecifierCapture = capture;
            }
        }

        return {
            referenceName: importSpecifierCapture.text,
            referenceSuffix:
                importSpecifierCapture.text.split(this.getNamespaceDelimiter()).pop() ?? "",
            sourceOfUsing: filePath,
            alias,
            source: filePath,
            usageType: "usage",
        };
    }

    private getGroupedImports(parsedFile: ParsedFile): ImportReference[] {
        const { filePath, language, tree } = parsedFile;

        const queryBuilder = new QueryBuilder(language);
        queryBuilder.setStatements([new SimpleQueryStatement(this.getGroupedImportsQuery())]);

        const groupedImportsQuery = queryBuilder.build();
        let importCaptures: QueryCapture[] = [];
        if (groupedImportsQuery !== undefined) {
            importCaptures = groupedImportsQuery.captures(tree.rootNode);
        }

        const importTextCaptures = getNameAndTextFromCaptures(importCaptures);

        dlog(importTextCaptures.toString());

        const importsOfFile: ImportReference[] = [];

        for (let index = 0; index < importTextCaptures.length; index++) {
            if (importTextCaptures[index].name === "namespace_use_alias_suffix") {
                // Todo: rewrite using at() method
                const matchingUsage = importsOfFile[importsOfFile.length - 1]; // eslint-disable-line unicorn/prefer-at
                // Split alias from alias keyword (if any) by space and use last element by pop()
                // it seems to be not possible to query the alias part only
                const alias = importTextCaptures[index].text.split(" ").pop() ?? "";

                this.fileToImportsBySuffixOrAliasMap
                    .get(filePath)
                    ?.delete(matchingUsage.referenceSuffix);
                if (alias.length > 0) {
                    matchingUsage.alias = alias;
                    this.fileToImportsBySuffixOrAliasMap.get(filePath)?.set(alias, matchingUsage);
                }

                continue;
            }

            const namespaceName = importTextCaptures[index].text;

            let hasUseGroupItem = importTextCaptures[index + 1]?.name === "namespace_use_item_name";
            let groupItemIndex = index;

            while (hasUseGroupItem) {
                const nextUseItem = importTextCaptures[groupItemIndex + 1];

                const usedNamespace =
                    namespaceName + this.getNamespaceDelimiter() + nextUseItem.text;

                const namespaceSuffix = nextUseItem.text;
                const importReferences: ImportReference = {
                    referenceName: usedNamespace,
                    referenceSuffix: namespaceSuffix,
                    sourceOfUsing: filePath,
                    alias: "",
                    source: filePath,
                    usageType: "usage",
                };

                importsOfFile.push(importReferences);
                this.fileToImportsBySuffixOrAliasMap
                    .get(filePath)
                    ?.set(importReferences.referenceSuffix, importReferences);

                hasUseGroupItem =
                    importTextCaptures[groupItemIndex + 2]?.name === "namespace_use_item_name";
                groupItemIndex++;
                index++;
            }
        }

        return importsOfFile;
    }

    private getUsages(
        parsedFile: ParsedFile,
        typesFromFile: Map<string, TypeInfo>,
        importReferences: ImportReference[],
    ): {
        candidates: TypeUsageCandidate[];
        unresolvedCallExpressions: UnresolvedCallExpression[];
    } {
        const { filePath, language, tree } = parsedFile;

        const queryBuilder = new QueryBuilder(language);
        queryBuilder.setStatements([new SimpleQueryStatement(this.getUsagesQuery())]);

        const usagesQuery = queryBuilder.build();
        let usagesCaptures: QueryCapture[] = [];
        if (usagesQuery !== undefined) {
            usagesCaptures = usagesQuery.captures(tree.rootNode);
        }

        const usagesTextCaptures: Array<{
            name: string;
            text: string;
            source: undefined | string;
            usageType: undefined | UsageType;
        }> = usagesCaptures.map((capture) => {
            return {
                name: capture.name,
                text: capture.node.text,
                usageType: undefined,
                source: undefined,
            };
        });

        dlog("class/object usages", usagesTextCaptures);

        const usagesAndCandidates: TypeUsageCandidate[] = [];
        const unresolvedCallExpressions: UnresolvedCallExpression[] = [];

        // Add implemented and extended classes as usages
        // to consider the coupling of those
        for (const [FQTN, FQTNInfo] of typesFromFile) {
            if (FQTNInfo.implementedFrom.length > 0) {
                for (const implementedClass of FQTNInfo.implementedFrom) {
                    usagesTextCaptures.push({
                        name: "qualified_name",
                        text: implementedClass,
                        usageType: "implements",
                        source: FQTN,
                    });
                }
            }

            if (FQTNInfo.extendedFrom !== undefined) {
                usagesTextCaptures.push({
                    name: "qualified_name",
                    text: FQTNInfo.extendedFrom,
                    usageType: "extends",
                    source: FQTN,
                });
            }
        }

        // Resolve usages against import statements and build concrete usages or usage candidates

        const processedQualifiedNames = new Set<string>();
        for (const usagesTextCapture of usagesTextCaptures) {
            const { name, usageType, source } = usagesTextCapture;
            let qualifiedName = usagesTextCapture.text;

            if (
                processedQualifiedNames.has(qualifiedName) ||
                (name !== "qualified_name" && name !== "call_expression")
            ) {
                continue;
            }

            processedQualifiedNames.add(qualifiedName);

            const qualifiedNameParts = qualifiedName.split(this.getNamespaceDelimiter());
            const cleanNameParts = qualifiedNameParts.map((namePart) => {
                if (namePart.endsWith("[]") || namePart.endsWith("()")) {
                    return namePart.slice(0, Math.max(0, namePart.length - 2));
                }

                if (namePart.endsWith("?")) {
                    return namePart.slice(0, Math.max(0, namePart.length - 1));
                }

                return namePart;
            });

            qualifiedName = cleanNameParts.join(this.getNamespaceDelimiter());

            const qualifiedNamePrefix = cleanNameParts.shift();
            if (qualifiedNamePrefix === undefined) {
                continue;
            }

            const resolvedImport = this.fileToImportsBySuffixOrAliasMap
                .get(filePath)
                ?.get(qualifiedNamePrefix);

            const FQTNInfos = typesFromFile.values();
            let fromFQTN = FQTNInfos.next().value as TypeInfo | undefined;
            if (!fromFQTN) {
                // No namespace found in current file
                break;
            }

            // Resolve the right entity/class/namespace in a file with multiple ones
            if (source !== undefined) {
                for (const [temporaryKey, temporaryValue] of typesFromFile.entries()) {
                    if (temporaryKey === source) {
                        fromFQTN = temporaryValue;
                    }
                }
            }

            const originalCleanNameParts = qualifiedName.split(this.getNamespaceDelimiter());

            if (resolvedImport === undefined) {
                if (
                    name === "call_expression" &&
                    this.getNamespaceDelimiter() === this.getFunctionCallDelimiter() &&
                    originalCleanNameParts.length > 1
                ) {
                    // Pop name of called function or similar callables from qualified name
                    originalCleanNameParts.pop();
                }

                const cleanQualifiedName = originalCleanNameParts.join(
                    this.getNamespaceDelimiter(),
                );
                processedQualifiedNames.add(cleanQualifiedName);

                // Skip current one if invalid space is included in potential class or namespace name
                if (cleanQualifiedName.includes(" ") || cleanQualifiedName.includes("<")) {
                    continue;
                }

                if (name === "call_expression") {
                    unresolvedCallExpressions.push(this.buildCallExpression(qualifiedName));
                }

                // For languages that allow the usage of classes in the same namespace without the need of an import:
                // add candidate in same namespace for unresolvable usage

                let usedNamespace = fromFQTN.namespace;
                if (!cleanQualifiedName.startsWith(fromFQTN.namespaceDelimiter)) {
                    usedNamespace += fromFQTN.namespaceDelimiter;
                }

                usedNamespace += cleanQualifiedName;

                // For languages that allow the usage of classes in parent namespace without the need of an import:
                // Look in parent namespaces for the used class name even if no import is present

                const parentNamespaceCandidates: string[] = [];
                if (fromFQTN.namespace.includes(this.getNamespaceDelimiter())) {
                    const sourceNamespaceParts = fromFQTN.namespace.split(
                        this.getNamespaceDelimiter(),
                    );
                    while (sourceNamespaceParts.length > 1) {
                        sourceNamespaceParts.pop();

                        const candidate =
                            sourceNamespaceParts.join(this.getNamespaceDelimiter()) +
                            this.getNamespaceDelimiter() +
                            cleanQualifiedName;
                        parentNamespaceCandidates.push(candidate);
                    }
                }

                // Heavy candidate building
                // combine current usage with all of the imports
                const moreCandidates: string[] = [];
                if (this.indirectNamespaceReferencing()) {
                    for (const importReference of importReferences) {
                        if (this.getNamespaceDelimiter() === this.getFunctionCallDelimiter()) {
                            const clonedNameParts = [...originalCleanNameParts];
                            while (clonedNameParts.length > 0) {
                                moreCandidates.push(
                                    importReference.referenceName +
                                        fromFQTN.namespaceDelimiter +
                                        clonedNameParts.join(this.getNamespaceDelimiter()),
                                );
                                clonedNameParts.pop();
                            }
                        } else {
                            moreCandidates.push(
                                importReference.referenceName +
                                    fromFQTN.namespaceDelimiter +
                                    cleanQualifiedName,
                            );
                        }
                    }
                }

                // Also, add candidate with exact used namespace
                // in the case that it is a fully qualified (root) namespace

                const finalCandidates = [
                    usedNamespace,
                    cleanQualifiedName,
                    ...moreCandidates,
                    ...parentNamespaceCandidates,
                ];
                for (const candidateUsedNamespace of finalCandidates) {
                    const usageCandidate: TypeUsageCandidate = {
                        usedNamespace: candidateUsedNamespace,
                        fromNamespace:
                            fromFQTN.namespace + fromFQTN.namespaceDelimiter + fromFQTN.typeName,
                        sourceOfUsing: filePath,
                        usageType: usageType ?? "usage",
                    };
                    // TODO prevent duplicate adds in current file
                    //  when is it a duplicate? (usedNamespace + fromNamespace + sourceOfUsing?)
                    usagesAndCandidates.push(usageCandidate);
                }
            } else {
                if (
                    name === "call_expression" &&
                    this.getNamespaceDelimiter() === this.getFunctionCallDelimiter()
                ) {
                    // Pop name of called function or similar callables
                    if (originalCleanNameParts.length > 1) {
                        cleanNameParts.pop();
                    }

                    // In case it cannot be resolved by resolvedImport name, try it later again
                    unresolvedCallExpressions.push(this.buildCallExpression(qualifiedName));
                }

                const modifiedQualifiedName = cleanNameParts.join(this.getNamespaceDelimiter());
                processedQualifiedNames.add(modifiedQualifiedName);

                // Skip current one if invalid space is included in potential class or namespace name
                if (modifiedQualifiedName.includes(" ") || modifiedQualifiedName.includes("<")) {
                    continue;
                }

                const usageCandidate: TypeUsageCandidate = {
                    usedNamespace:
                        resolvedImport.referenceName +
                        (cleanNameParts.length > 0
                            ? this.getNamespaceDelimiter() + modifiedQualifiedName
                            : ""),
                    fromNamespace:
                        fromFQTN.namespace + fromFQTN.namespaceDelimiter + fromFQTN.typeName,
                    sourceOfUsing: filePath,
                    usageType: usageType ?? "usage",
                };
                usagesAndCandidates.push(usageCandidate);

                if (
                    resolvedImport.alias !== "" &&
                    this.noImportForClassesInSameOrParentNamespaces()
                ) {
                    // In This case, alias can be a Qualified Name
                    // Add Same Namespace Candidate
                    // Add Parent Namespace Candidate
                    const fromNamespaceParts = fromFQTN.namespace.split(
                        this.getNamespaceDelimiter(),
                    );
                    while (fromNamespaceParts.length > 0) {
                        const usageCandidate: TypeUsageCandidate = {
                            usedNamespace:
                                fromNamespaceParts.join(this.getNamespaceDelimiter()) +
                                fromFQTN.namespaceDelimiter +
                                resolvedImport.referenceName,
                            fromNamespace:
                                fromFQTN.namespace +
                                fromFQTN.namespaceDelimiter +
                                fromFQTN.typeName,
                            sourceOfUsing: filePath,
                            usageType: usageType ?? "usage",
                        };
                        usagesAndCandidates.push(usageCandidate);
                        fromNamespaceParts.pop();
                    }
                }
            }
        }

        return {
            candidates: usagesAndCandidates,
            unresolvedCallExpressions,
        };
    }

    private buildCallExpression(qualifiedName: string): {
        name: string;
        variableNameIncluded: boolean;
        namespaceDelimiter: string;
    } {
        return {
            name: qualifiedName,
            variableNameIncluded:
                this.getNamespaceDelimiter() === this.getFunctionCallDelimiter() &&
                qualifiedName.includes(this.getNamespaceDelimiter()),
            namespaceDelimiter: this.getNamespaceDelimiter(),
        };
    }
}
