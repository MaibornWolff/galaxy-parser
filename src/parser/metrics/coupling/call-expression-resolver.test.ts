import { describe, expect, it } from "vitest";
import { type Relationship } from "../metric.js";
import { type UnresolvedCallExpression } from "../../resolver/call-expressions/abstract-collector.js";
import { type Accessor } from "../../resolver/accessors/abstract-collector.js";
import { getAdditionalRelationships } from "./call-expression-resolver.js";

describe("CallExpressionResolver", () => {
    describe("resolves call expressions and retrieves additional and transitive relationships", () => {
        it("when call expressions with save calls, public accessors and the right dependencies are given", () => {
            const firstItem: Relationship = {
                fromNamespace: "FirstItemNamespace.FirstItem",
                fromSource: "FirstItem",
                toSource: "SecondItem",
                toNamespace: "SecondItemNamespace.SecondItem",
                fromClassName: "FirstItem",
                toClassName: "SecondItem",
                usageType: "usage",
            };

            const secondItem: Relationship = {
                fromNamespace: "SecondItemNamespace.SecondItem",
                fromSource: "SecondItem",
                toSource: "ThirdItem",
                toNamespace: "ThirdItemNamespace.ThirdItem",
                fromClassName: "FirstItem",
                toClassName: "ThirdItem",
                usageType: "usage",
            };

            const thirdItem: Relationship = {
                fromNamespace: "ThirdItemNamespace.ThirdItem",
                fromSource: "ThirdItem",
                toSource: "FourthItem",
                toNamespace: "FourthItemNamespace.FourthItem",
                fromClassName: "FirstItem",
                toClassName: "FourthItem",
                usageType: "usage",
            };

            const dependencyTree = new Map<string, Relationship[]>();
            dependencyTree.set(firstItem.fromSource, [firstItem]);
            dependencyTree.set(secondItem.fromSource, [secondItem]);
            dependencyTree.set(thirdItem.fromSource, [thirdItem]);
            dependencyTree.set(thirdItem.toSource, []);

            const callExpression1: UnresolvedCallExpression = {
                name: "myVariable?.UnknownAccessor?.AccessorInSecondItem?.AccessorInThirdItem?",
                namespaceDelimiter: ".",
                variableNameIncluded: true,
            };

            const callExpression2: UnresolvedCallExpression = {
                name: "myVariable.AccessorInSecondItem",
                namespaceDelimiter: ".",
                variableNameIncluded: true,
            };

            const unresolvedCallExpressions = new Map<string, UnresolvedCallExpression[]>();
            unresolvedCallExpressions.set(firstItem.fromSource, [callExpression1, callExpression2]);

            const accessor1: Accessor = {
                filePath: "SecondItem",
                name: "AccessorInSecondItem",
                FQTNInfos: [
                    {
                        namespace: "SecondItemNamespace",
                        sourceFile: "SecondItem",
                        typeName: "SecondItem",
                        classType: "class",
                        namespaceDelimiter: ".",
                        implementedFrom: [],
                    },
                ],
                returnType: "ThirdItem",
            };

            const accessor2: Accessor = {
                filePath: "ThirdItem",
                name: "AccessorInThirdItem",
                FQTNInfos: [
                    {
                        namespace: "ThirdItemNamespace",
                        sourceFile: "ThirdItem",
                        typeName: "ThirdItem",
                        classType: "class",
                        namespaceDelimiter: ".",
                        implementedFrom: [],
                    },
                ],
                returnType: "FourthItem",
            };

            const publicAccessors = new Map<string, Accessor[]>();
            publicAccessors.set(accessor1.name, [accessor1]);
            publicAccessors.set(accessor2.name, [accessor2]);

            const additionalRelationships = getAdditionalRelationships(
                dependencyTree,
                unresolvedCallExpressions,
                publicAccessors,
                new Set<string>(),
            );

            expect(additionalRelationships).toMatchSnapshot();
        });
    });
});
