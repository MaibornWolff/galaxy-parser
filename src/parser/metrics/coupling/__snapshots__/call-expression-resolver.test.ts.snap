// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CallExpressionResolver > resolves call expressions and retrieves additional and transitive relationships > when call expressions with save calls, public accessors and the right dependencies are given 1`] = `
[
  {
    "fromFQTN": "FirstItemNamespace.FirstItem",
    "fromFile": "FirstItem",
    "fromTypeName": "ThirdItem",
    "toFQTN": "ThirdItemNamespace.ThirdItem",
    "toFile": "ThirdItem",
    "toTypeName": "ThirdItem",
    "usageType": "usage",
  },
  {
    "fromFQTN": "FirstItemNamespace.FirstItem",
    "fromFile": "FirstItem",
    "fromTypeName": "FourthItem",
    "toFQTN": "FourthItemNamespace.FourthItem",
    "toFile": "FourthItem",
    "toTypeName": "FourthItem",
    "usageType": "usage",
  },
]
`;
